kymatio/scattering2d/backend/backend_skcuda.py

```diff
@@ -265,32 +269,37 @@ def cdgmm(A, B, inplace=False):
             C[b, c, m, n, :] = A[b, c, m, n, :] * B[m, n, :]
     """
     A, B = A.contiguous(), B.contiguous()
-    if A.size()[-3:] != B.size():
+    if A.size()[-3:-1] != B.size()[-3:-1]:
         raise RuntimeError('The filters are not compatible for multiplication!')
 
-    if not iscomplex(A) or not iscomplex(B):
-        raise TypeError('The input, filter and output should be complex')
-
-    if B.ndimension() != 3:
-        raise RuntimeError('The filters must be simply a complex array!')
-
-    if type(A) is not type(B):
-        raise RuntimeError('A and B should be same type!')
-
-    if not A.is_cuda:
-        raise RuntimeError('Use the torch backend for cpu tensors!')
+    if not iscomplex(A):
+        raise TypeError('The input must be complex')
 
-    C = A.new(A.size()) if not inplace else A
-    m, n = B.nelement() // 2, A.nelement() // B.nelement()
-    lda = m
-    ldc = m
-    incx = 1
-    handle = torch.cuda.current_blas_handle()
-    stream = torch.cuda.current_stream()._as_parameter_
-    cublas.cublasSetStream(handle, stream)
-    cublas.cublasCdgmm(handle, 'l', m, n, A.data_ptr(), lda, B.data_ptr(), incx, C.data_ptr(), ldc)
-    return C
+    if not iscomplex(B) and not isreal(B):
+        raise TypeError('The filter must be complex or real')
 
+    if B.ndimension() != 3:
+        raise RuntimeError('The filter must be a 3-tensor')
 
+    if A.dtype is not B.dtype:
+        raise RuntimeError('A and B must be of the same dtype')
 
+    if A.device != B.device:
+        raise RuntimeError('A and B must be on the same device')
 
+    if isreal(B):
+        if inplace:
+            return A.mul_(B)
+        else:
+            return A * B
```

Comment:
though technically, since this is the skcuda backend, we can't really expect such high-level ideas to work here.

I'm fine with adding this here for the moment even though it is incongruent, because we have already done this with the fft modules.

When we start adding other front/backends, we might need to revisit this and have it use the `dgmm` function (could be faster anyway if pytorch isn't smart at chunking/reshaping due to the implicit broadcast)

Reply:
> since this is the skcuda backend, we can't really expect such high-level ideas to work here.

But the skcuda backend is still pytorch. There's no need to go low-level. At least for now, the skcuda backend means "you *also* need scikit-cuda", not "you need scikit-cuda instead of pytorch". Is this planned to change?

> `dgmm` function (could be faster anyway

That would be an argument. (But then pytorch should be using it as well.)